example/tagSpecialization/src/tagSpecialization.cpp

```diff
@@ -0,0 +1,136 @@
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
+ * IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <alpaka/alpaka.hpp>
+#include <alpaka/example/ExampleDefaultAcc.hpp>
+
+#include <iostream>
+
+#if 0
+//! Kernel to illustrate specialization for a particular accelerator
+//!
+//! It has a generic operator() implementation and an overload for the CUDA accelerator.
+//! When running the kernel on a CUDA device, the corresponding overload of operator() is called.
+//! Otherwise the generic version is called.
+//! The same technique can be applied for any function called from inside the kernel,
+//! thus allowing specialization of only relevant part of the code.
+//! It can be useful for optimization or accessing specific functionality not abstracted by alpaka.
+//!
+//! This kernel demonstrates the simplest way to achieve the effect by function overloading.
+//! Note that it does not perform specialization (in C++ template meaning) of function templates.
+//! We use the word "specialization" as it represents a case of having a special version for a particular accelerator.
+//! One could apply a similar technique by having an additional class template parametrized by the accelerator type.
+//! For such a case, both template specialization and function overloading of the methods can be employed.
+struct Kernel
+{
+    //! Implementation for the general case
+    //!
+    //! It will be called when no overload is a better match.
+    template<typename TAcc>
+    ALPAKA_FN_ACC auto operator()(TAcc const& acc) const
+    {
+        // For simplicity assume 1d thread indexing
+        auto const globalThreadIdx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0];
+        if(globalThreadIdx == 0u)
+            printf("Running the general kernel implementation\n");
+    }
+
+    //! Simple overload to have a special version for the CUDA accelerator
+    //!
+    //! We have to guard it with #ifdef as the types of alpaka accelerators are only conditionally available.
+    //! Overloading for other accelerators is similar, with another template name instead of AccGpuCudaRt.
+#    ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
+    template<typename TDim, typename TIdx>
+    ALPAKA_FN_ACC auto operator()(alpaka::AccGpuCudaRt<TDim, TIdx> const& acc) const
+    {
+        // This overload is used when the kernel is run on the CUDA accelerator.
+        // So inside we can use both alpaka and native CUDA directly.
+        // For simplicity assume 1d thread indexing
+        auto const globalThreadIdx = blockIdx.x * gridDim.x + threadIdx.x;
+        if(globalThreadIdx == 0)
+            printf("Running the specialization for the CUDA accelerator\n");
+    }
+#    endif
+};
+
+#endif
+
+template<typename TAcc>
+std::string host_function()
+{
+    return "generic host function";
+}
+
+template<>
+std::string host_function<alpaka::AccOaccTag>()
+{
+    return "OpenAcc host function";
+}
```

Comment:
I tend to avoid full-specialization for function templates, because I have been told so by me mentor years ago. I would rather use a parameter and overloading:
```c++
template<typename Tag>
std::string host_function(Tag)
{
    return "generic host function";
}

std::string host_function(alpaka::AccOaccTag)
{
    return "OpenAcc host function";
}
```
Maybe you can add this as an additional example!

Reply:
done